pkg/apis/tenancy/v1alpha1/types.go

```diff
@@ -210,15 +225,48 @@ type ClusterWorkspaceTypeSpec struct {
 	AllowedParentWorkspaceTypes []ClusterWorkspaceTypeName `json:"allowedParentWorkspaceTypes,omitempty"`
 }
 
+// ClusterWorkspaceTypeExtension defines how other ClusterWorkspaceTypes are
+// composed together to add functionality to the owning ClusterWorkspaceType.
+type ClusterWorkspaceTypeExtension struct {
+	// with are ClusterWorkspaceTypes whose initializers are added to the list
+	// for the owning type, and for whom the owning type becomes an alias.
+	//
+	// +optional
+	With []ClusterWorkspaceTypeReference `json:"with,omitempty"`
```

Comment:
am no native speaker, "extend with" sounds odd. Don't I "extend X, without Y" ?

Reply:
What is worth, and I think this is a blocker: "with" and "without" suggests these to be inverse, anti-symmetric. And they are clearly not. Without only applies one level deep, i.e. removes one edge in your graph language. Another edge or actually path in the graph will replace that edge:

A extends B without X.
B extends X,Y.
Y extends X.

Result: A extends X,Y.

Reply:
Actually, I realized that there is something else unexpected:

I don't think `with []reference, without []reference` is a good choice. That's partly the reason for the inconsistent semantics. We rather want `[]{with: reference, without []reference}`.

Reply:
The reason why it is not a good choice: it is not continuous if a dependee changes later-on. Suddenly, we have a broken invariant when somebody adds another inheritance. That's an unfortunate API:

Example: A extends with B,C, without X.

Assume yesterday neither B not C extend anything. Then today B extends X. Because we exclude X, A cannot be an alias for B, while it was an alias yesterday. I.e. adding X to B leads to a negative effect today. Non-positive effects are bad for composability and UX.

Reply:
> A extends B without X.
B extends X,Y.
Y extends X.

> Result: A aliases X,Y (not B!)

This is not right ... it's resolved at each node so A aliases only Y

Reply:
Fixing in upcoming push